cli/internal/oauth/manager/manager.go

```diff
@@ -0,0 +1,150 @@
+package manager
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/cli/cli/internal/oauth/api"
+	"github.com/docker/cli/cli/oauth"
+
+	"github.com/pkg/browser"
+)
+
+// OAuthManager is the manager
+type OAuthManager struct {
+	api         api.OAuthAPI
+	audience    string
+	tenant      string
+	openBrowser func(string) error
+}
+
+// OAuthManagerOptions are the options used for New to create a new auth manager.
+type OAuthManagerOptions struct {
+	Audience    string
+	ClientID    string
+	Scopes      []string
+	Tenant      string
+	DeviceName  string
+	OpenBrowser func(string) error
+}
+
+func New(options OAuthManagerOptions) *OAuthManager {
+	scopes := []string{"openid", "offline_access"}
+	if len(options.Scopes) > 0 {
+		scopes = options.Scopes
+	}
+
+	openBrowser := browser.OpenURL
+	if options.OpenBrowser != nil {
+		openBrowser = options.OpenBrowser
+	}
+
+	return &OAuthManager{
+		audience: options.Audience,
+		api: api.API{
+			BaseURL:  "https://" + options.Tenant,
+			ClientID: options.ClientID,
+			Scopes:   scopes,
+		},
+		tenant:      options.Tenant,
+		openBrowser: openBrowser,
+	}
+}
+
+// LoginDevice launches the device authentication flow with the tenant, printing
+// instructions to the provided writer and attempting to open the browser for
+// the user to authenticate.
+// Once complete, the retrieved tokens are returned.
+func (m *OAuthManager) LoginDevice(ctx context.Context, w io.Writer) (*oauth.TokenResult, error) {
+	state, err := m.api.GetDeviceCode(ctx, m.audience)
+	if err != nil {
+		return nil, fmt.Errorf("login failed: %w", err)
+	}
+
+	if state.UserCode == "" {
+		return nil, errors.New("login failed: no user code returned")
+	}
+
+	_, _ = fmt.Fprintln(w, "\nYou will be signed in using a web-based login.")
+	_, _ = fmt.Fprintln(w, "To sign in with credentials on the command line, use 'docker login -u <username>'")
+	_, _ = fmt.Fprintf(w, "\nYour one-time device confirmation code is: %s\n", state.UserCode)
+	_, _ = fmt.Fprint(w, "\nPress ENTER to open the browser.\n")
+	_, _ = fmt.Fprintf(w, "Or open the URL manually: %s\n", strings.Split(state.VerificationURI, "?")[0])
```

Comment:
What do you think about this:

```suggestion
	_, _ = fmt.Fprintf(w, "Submit your device code here: %s\n", strings.Split(state.VerificationURI, "?")[0])
	_, _ = fmt.Fprint(w, "\n\033[1mPress ENTER\033[0m to open your browser…\n")
```

Reply:
![Screenshot 2024-08-05 at 13 51 22](https://github.com/user-attachments/assets/dd0b3979-9964-4630-b1f0-b2c7d23dd671)

I wonder if this is too many ellipsis(es) now 😅

Reply:
Oh right, there was one already. 😅 Agree that we should only have one. Let's skip the one I added.

Reply:
But let's make it an actual ellipsis :D

Reply:
Sounds good!